### PR TITLE
HIVE-113267: Create order and link tasks for surgery related instructions

### PR DESCRIPTION
**Pull request by vvkpd**
## Summary

- **`SurgeryOrderPostSaveCommandImpl`** — post-save command creating Surgery/General Order per form submission. Order lookup keyed on `surgicalAppointmentUuid` ensuring different surgical appointments in the same encounter each get their own Surgery Order.
- **`SurgeryObsOrderLinkDao`** — new DAO (`assignOrderToUnlinkedObs`) in `bahmni-emr-api` layer extracts HQL bulk update from command.
- **`SurgicalBlockObsHandler`** — `ComplexObsHandler` for the surgery selection concept.
- **Liquibase** — seeds `Surgery Order` and `General Order` order types.
- **`pom.xml`** — `operationtheater-api` version managed via `${operationtheater.version}` property in parent pom.

## Feature Toggle via Global Property

```
bahmnicore.order.surgerySelectionConceptUuid = <UUID of Select Surgery concept>
```

- **GP configured** → concept found → Surgery/General Order created for all form submissions ✅
- **GP not configured** → concept null → order creation skipped gracefully, no errors ✅

This makes the feature **incrementally deployable** — ship the code, configure the GP when ready. No flag days, no breaking changes.

The GP value (concept UUID) is seeded via `cure-bahmni-emr` master data — not owned by bahmni-core since the concept is deployment-specific.

## Behaviour

| Scenario | Result |
|---|---|
| Surgery form (Select Surgery filled) | `Surgery Order` created; obs + `surgical_appointment.order_id` linked |
| Non-surgery form (same encounter) | `General Order` created; obs linked separately |
| Same surgical appointment submits again | Existing Surgery Order reused (keyed on appointment UUID) |
| Different surgical appointment, same encounter | New Surgery Order — does NOT reuse the first surgery's order |
| Form re-edit | No duplicate order; only newly added obs linked (`WHERE order IS NULL`) |
| GP not configured | No order created; encounter save unaffected |

## Review Comments Addressed

| Comment | Status |
|---|---|
| `getConceptByName()` null guard | ✅ Fixed |
| `careSetting` null guard | ✅ Fixed |
| `ProviderService` unused | ✅ Fixed — removed |
| `SessionFactory` → DAO | ✅ Fixed — `SurgeryObsOrderLinkDao.assignOrderToUnlinkedObs()` |
| `operationtheater-api` version hardcoded | ✅ Fixed — `${operationtheater.version}` |
| `UUID()` MySQL-specific in Liquibase | ✅ Fixed — fixed UUID literals |
| `SUPPORTED_VIEWS` naming | ✅ Fixed — renamed to `supportedViews` |
| Concept hardcoded | ✅ Fixed — configurable via `bahmnicore.order.surgerySelectionConceptUuid` GP |

## Open Items (Pending Tech Lead Discussion)

| Item | Notes |
|---|---|
| `require_module operationtheater` in `config.xml` | Cross-module integration decision — see `HIVE-113267_cross_module_integration.md` |
| Concept for General Order | "Select Surgery" used as DB placeholder for both order types — acceptable since no concept class constraints. Option B (dedicated concept) deferred. |

## Follow-up PRs

- `openmrs-module-operationtheater` [#1](https://github.com/cureinternational/openmrs-module-operationtheater/pull/1)
- `openmrs-module-fhir2Extension` [#4](https://github.com/cureinternational/openmrs-module-fhir2Extension/pull/4)
- `openmrs-module-ipd-frontend` [#77](https://github.com/cureinternational/openmrs-module-ipd-frontend/pull/77)
- `openmrs-module-bahmniapps` [#174](https://github.com/cureinternational/openmrs-module-bahmniapps/pull/174)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        Provider provider = getPrimaryProvider(encounter);
+        if (provider == null) {
+            return null;
+        }
+
+        Order order = new Order();
+        order.setPatient(encounter.getPatient());
+        order.setEncounter(encounter);
+        order.setOrderType(orderType);
+        order.setConcept(conceptService.getConceptByName(SELECT_SURGERY_CONCEPT_NAME));
```
Comment (vvkpd) [inline bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java]: blocking: `conceptService.getConceptByName(SELECT_SURGERY_CONCEPT_NAME)` can return null if 'Select Surgery' is not seeded (or has a different locale/name in this env). `orders.concept_id` is NOT NULL at the DB level, so `saveOrder` will throw a constraint violation, rolling back the entire encounter save.

Also: this uses the 'Select Surgery' concept for both `Surgery Order` AND `General Order`. For a non-surgery form, using the surgery-selection concept is semantically incorrect.

```java
Concept orderConcept = conceptService.getConceptByName(SELECT_SURGERY_CONCEPT_NAME);
if (orderConcept == null) {
    return null; // concept not seeded
}
order.setConcept(orderConcept);
```

File: bahmnicore-omod/src/main/resources/config.xml [hunk 2]
```diff
@@ -34,6 +34,7 @@
         <require_module>org.openmrs.module.auditlog</require_module>
         <require_module>org.bahmni.module.bahmnicommons</require_module>
         <require_module>org.bahmni.module.communication</require_module>
+        <require_module>org.openmrs.module.operationtheater</require_module>
```
Comment (vvkpd) [inline bahmnicore-omod/src/main/resources/config.xml:37]: blocking: `require_module` makes `operationtheater` a hard runtime dependency — `bahmnicore-omod` will refuse to start on any deployment where the OT module is absent (vanilla Bahmni, non-surgical sites). The Maven `scope=provided` correctly signals optional compile-time availability, but `require_module` contradicts that at the OpenMRS runtime layer.

The `linkSurgicalAppointmentToOrder` path only fires when a 'Select Surgery' obs is present (CURE-specific), so non-OT deployments would never hit it anyway.

Remove this entry and guard the `Context.getService` call in `linkSurgicalAppointmentToOrder` with a module-started check.

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        order.setCareSetting(orderService.getCareSettingByName("OUTPATIENT"));
```
Comment (vvkpd) [inline bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java]: suggestion: `getCareSettingByName("OUTPATIENT")` can return null — `order.setCareSetting(null)` will cause `saveOrder` to throw. The same guard pattern used for `orderType` (lines 106-108) should apply here.

Also consider `CareSetting.CareSettingType.OUTPATIENT.toString()` instead of the bare string (pattern in `LabResultPersister` / `AccessionHelper`).

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    static final String SURGERY_ORDER_TYPE_NAME = "Surgery Order";
+    static final String GENERAL_ORDER_TYPE_NAME = "General Order";
+
+    private final OrderService orderService;
+    private final ConceptService conceptService;
+    private final ProviderService providerService;
+    private final SessionFactory sessionFactory;
+
+    @Autowired
+    public SurgeryOrderPostSaveCommandImpl(OrderService orderService, ConceptService conceptService,
```
Comment (vvkpd) [inline bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java:38]: suggestion: `ProviderService` is imported, declared as a field, and accepted in the constructor — but never called anywhere in the class. `getPrimaryProvider()` reads directly from `encounter.getEncounterProviders()`. Remove from import, field, constructor, and test mock.

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        for (EncounterProvider ep : encounter.getEncounterProviders()) {
+            if (ep.getProvider() != null) {
+                return ep.getProvider();
+            }
+        }
+        return null;
+    }
+
+    // HQL bulk update bypasses ImmutableEntityInterceptor (same pattern as BahmniObsDaoImpl.updateObsMember).
+    private void linkAllObsToOrder(Order order, Encounter encounter) {
```
Comment (vvkpd) [inline bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java]: suggestion: Raw HQL bulk-update methods belong in a DAO layer. The comment cites `BahmniObsDaoImpl` as precedent — but `BahmniObsDaoImpl` IS a DAO that encapsulates `SessionFactory`. Every other raw HQL in this project lives behind a `*DaoImpl`.

Consider extracting a `SurgeryObsOrderLinkDao` in `bahmnicore-api/.../dao/` and injecting the DAO instead of `SessionFactory` directly.

File: bahmni-emr-api/pom.xml [hunk 6]
```diff
@@ -167,6 +167,12 @@
             <type>test-jar</type>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.openmrs.module</groupId>
+            <artifactId>operationtheater-api</artifactId>
+            <version>1.9.0-SNAPSHOT</version>
```
Comment (vvkpd) [inline bahmni-emr-api/pom.xml]: suggestion: `1.9.0-SNAPSHOT` is hardcoded inline. All other module versions in this project use `${property}` from the parent `pom.xml` (e.g. `${emrapi-omod.version}`).

Add `<operationtheater.version>1.9.0-SNAPSHOT</operationtheater.version>` to parent `pom.xml` properties and reference as `${operationtheater.version}`.

File: bahmnicore-omod/src/main/resources/liquibase.xml [hunk 7]
```diff
@@ -4788,6 +4788,33 @@
...
+        <preConditions onFail="MARK_RAN">
+            <sqlCheck expectedResult="0">SELECT COUNT(*) FROM order_type WHERE name = 'Surgery Order'</sqlCheck>
+        </preConditions>
+        <insert tableName="order_type">
+            <column name="name" value="Surgery Order"/>
+            <column name="description" value="Order created when an operative report is linked to a surgical appointment"/>
+            <column name="java_class_name" value="org.openmrs.Order"/>
+            <column name="date_created" valueDate="now()"/>
+            <column name="creator" value="1"/>
+            <column name="uuid" valueComputed="UUID()"/>
```
Comment (vvkpd) [inline bahmnicore-omod/src/main/resources/liquibase.xml]: suggestion: `UUID()` is MySQL-specific. On PostgreSQL (which Bahmni supports) this will fail with "function uuid() does not exist", blocking the migration run.

Consider using a fixed UUID literal or a Liquibase-portable approach.

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+                .setParameter("order", order)
+                .setParameter("encounter", encounter)
+                .executeUpdate();
+    }
+
+    private void linkSurgicalAppointmentToOrder(String surgicalApptUuid, Order order) {
+        if (surgicalApptUuid == null) {
+            return;
+        }
+        SurgicalAppointmentService svc = Context.getService(SurgicalAppointmentService.class);
```
Comment (vvkpd) [inline bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java:163]: nit: Once `require_module` is removed, `Context.getService(SurgicalAppointmentService.class)` may return null if OT is not deployed. Add a null guard to prevent NPE:
```java
SurgicalAppointmentService svc = Context.getService(SurgicalAppointmentService.class);
if (svc == null) return;
```

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/obs/handler/SurgicalBlockObsHandler.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import org.openmrs.obs.ComplexObsHandler;
+import org.openmrs.obs.handler.AbstractHandler;
+import org.springframework.stereotype.Component;
+
+@Component
+public class SurgicalBlockObsHandler extends AbstractHandler implements ComplexObsHandler {
+
+    public static final Log log = LogFactory.getLog(SurgicalBlockObsHandler.class);
+
+    private static final String[] SUPPORTED_VIEWS = new String[] {
```
Comment (vvkpd) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/obs/handler/SurgicalBlockObsHandler.java]: nit: Sibling handlers in this package (`LocationObsHandler`, `ProviderObsHandler`) use `supportedViews` (camelCase). Rename `SUPPORTED_VIEWS` to `supportedViews` for consistency.

File: bahmnicore-omod/src/main/resources/liquibase.xml [hunk 10]
```diff
@@ -4788,6 +4788,33 @@
         </sql>
     </changeSet>
 
+    <changeSet id="cure-surgery-order-type-1" author="vivek">
```
Comment (vvkpd) [inline bahmnicore-omod/src/main/resources/liquibase.xml]: nit: Author `"vivek"` — the only other CURE-authored changeset in this file uses `"vivek-tw"`. Update for consistency.

**Review by vvkpd — COMMENTED** [review]
## Review for HIVE-113267

Solid implementation of the core order-linking flow — the post-save command pattern, idempotency logic, and HQL bulk update approach are all well thought out. Found 2 blocking issues and 6 suggestions to address before merge.

Found 2 blocking issue(s), 6 suggestion(s), and 3 nit(s).

**Acceptance Criteria**: 3/3 met (Scenario 2 surgical-appointment linkage depends on OT module being deployed — see blocking comment on config.xml)

### Issues not in diff

| File | Type | Issue |
|---|---|---|
| `SurgeryOrderPostSaveCommandImplTest.java` | nit | Missing test: `shouldNotCreateOrderWhenNoProviderOnEncounter` — no-provider case silently skips order creation with no log and no test |
| `SurgeryOrderPostSaveCommandImplTest.java` | nit | Missing test: `shouldNotSaveSurgicalAppointmentWhenNotFoundByUuid` — null appointment return should not call `svc.save()` |

File: bahmnicore-omod/src/main/resources/liquibase.xml [hunk 11]
```diff
@@ -4788,6 +4788,33 @@
...
     </changeSet>
 
+    <changeSet id="add-surgery-order-type" author="vivek">
+        <preConditions onFail="MARK_RAN">
+            <sqlCheck expectedResult="0">SELECT COUNT(*) FROM order_type WHERE name = 'Surgery Order'</sqlCheck>
+        </preConditions>
+        <insert tableName="order_type">
+            <column name="name" value="Surgery Order"/>
+            <column name="description" value="Order created when an operative report is linked to a surgical appointment"/>
+            <column name="java_class_name" value="org.openmrs.Order"/>
```
Comment (SooryaKumaranC-tw) [inline bahmnicore-omod/src/main/resources/liquibase.xml:4798]: Do we need this JavaClassName? As Anyways it is going to be Order only.

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+        }
+
+        surgeryObsOrderLinkDao.assignOrderToUnlinkedObs(newOrder, currentEncounter);
+        linkSurgicalAppointmentToOrder(surgicalApptUuid, newOrder);
+
+        return updatedEncounterTransaction;
+    }
+
+    private Concept getSurgerySelectionConcept() {
+        String conceptUuid = adminService.getGlobalProperty(SURGERY_SELECTION_CONCEPT_UUID_GP, "");
```
Comment (SooryaKumaranC-tw) [inline bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java:74]: I can't see the global property added in the EMR repo changes(https://github.com/cureinternational/cure-bahmni-emr/pull/673/changes) was it missed?

File: bahmnicore-omod/src/main/resources/liquibase.xml (hunk 11 above)
Comment (vvkpd) [reply bahmnicore-omod/src/main/resources/liquibase.xml:4798]: we are running insert cmd, the column is present so filling the right value.

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java (hunk 12 above)
Comment (vvkpd) [reply bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/command/impl/SurgeryOrderPostSaveCommandImpl.java:74]: Yes, I have to raise another PR for that. That is pending

File: bahmnicore-omod/src/main/resources/liquibase.xml (hunk 11 above)
Comment (SooryaKumaranC-tw) [reply bahmnicore-omod/src/main/resources/liquibase.xml:4798]: Okay

**Review by SooryaKumaranC-tw — APPROVED** [review]

**Review by sasikiran-tw — APPROVED** [review]

**Review by binduak — APPROVED** [review]
